### ISSUE TITLE
Metadata identifier concept

### ISSUE THREAD

**Issue by mih** [datalad/datalad-metalad#30]
### Dataset

- `@id` is the git commit of the dataset
- `identifier` is dataset ID

### File

- `@id` is the hash of dataset ID (not Git commit!) + relative path within the dataset + `contentid`

### Content

- `@id` is a git-annex key (or something like it for files in Git)
- if the file extension is problematic, users must not use the `*E` backends

**Comment by bpoldrack** [comment]
@mih: Re Subdataset:
> - @id is a hash of dataset id + relative path + subdataset commit

Unclear to me whether that's referring to the dataset's ID or `@id` - so: commit or not?
And apart from that, shouldn't "being a subdataset" be a relation? And if so, the relative path would be the only thing specific as both datasets are identified by their commit already, no?

To be clear, I'm not suggesting it's somehow wrong, I'm just not clear about the idea what needs what kind of identification and why. It might well be absolutely fine as is.

**Comment by tgbugs** [comment]
We didn't discuss this but subdataset `@id` should be computed from the `dataset identity + relative path + subdataset commit` using the dataset `@id` would cause the subdataset identity to change spuriously when files in the parent dataset changed.

**Comment by bpoldrack** [comment]
@tgbugs: True. But then: A dataset is not a subdataset of another dataset independently of the commit of that superdataset. It is introduced as a subdatset by a particular commit, and it can be (re-)moved. So, what would be the notion of "subdataset" that would be identified by this?

**Comment by tgbugs** [comment]
I am equally curious about this and I left out a sentence asking exactly the same question. I don't know what use cases @mih envisions.

**Comment by mih** [comment]
Thx for your comments.

> We didn't discuss this but subdataset `@id` should be computed from the `dataset identity + relative path + subdataset commit` using the dataset `@id` would cause the subdataset identity to change spuriously when files in the parent dataset changed.

This is exactly what I had in mind.

> @tgbugs: True. But then: A dataset is not a subdataset of another dataset independently of the commit of that superdataset. It is introduced as a subdatset by a particular commit, and it can be (re-)moved. So, what would be the notion of "subdataset" that would be identified by this?

My current concept is this (quick sketch):

```
    {
      # <A> dataset commit
      "@id": "af0b335b73e0952c216e57d2b1aeb06d11780bae",
      # <B> random identifier, doesn't change with commits
      "identifier": "e420a0fa-cba8-11e9-87f3-f0d5bf7b5561",
      "@type": "Dataset",
      "hasPart": [
        { "@type": "Dataset",
          # <C>
          "name": <relative path within dataset>
          "@id": "hash(<B> + <C> + <subdataset commit>)",
          "sameAs": <subdataset commit>
 ...
```
This setup should have the desired features. The subdataset relationship is declared within the context of the superdataset and a specific version/commit. The document describing the dataset in the subdataset context has a dedicated `@id` for this scope. The `sameAs` in the `hasPart` record is needed, because the subdataset's metadata is extracted from that dataset itself, which is not aware (and should not be) of its nature of a subdataset in that specific context (it could be a subdataset of any number of other datasets, at its specific version/commit).

Technically, it will be a bit tricky to figure out which specific commit to use as `<subdataset commit>`, because we need to avoid spurious changes when DataLad's metadata aggregation alone causes a commit to be made, i.e. there is absolutely no change to the data the metadata is about. I will have to think about this more, but I don't think there will be a fundamental challenge.

**Comment by bpoldrack** [comment]
Okay. Now it makes sense to me ;-)
Minor remark on field names: 

> `"name": <relative path within dataset>`

Git submodules in theory have a name as well as a path (although identical by default). It might be better to call that field "path" rather than "name" to avoid potential confusion/conflict in the future.

**Comment by mih** [comment]
> It might be better to call that field "path" rather than "name" to avoid potential confusion/conflict in the future

True, didn't want to add this complexity to this issue, though. Within the actual metadata we should have both. In general, I am very much at a loss how to optimally annotate and filesystem-related aspects. See https://github.com/datalad/datalad-metalad/issues/1

**Comment by tgbugs** [comment]
This all seems consistent with how you plan to deal with files, so I think it will be ok.

**Comment by yarikoptic** [comment]
New to the discussion here. It feels that there is a lose end for @id is at the top: identify a dataset uniquely. Unlike for all other IDs, which pretty much incorporate their parent structure information, the same isn't done for dataset itself. I can have two instances of a dataset locally, cannot identify them uniquely. Might be just needing clarification, that the goal isn't to identify any particular "instance", but rather "substance" ;-)

**Comment by yarikoptic** [comment]
Or might be worth adding an explicit layer on top like Storage, or Filesystem for a more concrete sample (could be be website, portal etc), which would then follow the same logic to reference datasets, and hopefully relying on already established ways to Id them (eg filesystem blkid etc). And then be

- identify a particular dataset provided at specific location and version
- @id is a hash of stage id + relative path + dataset id

This could provide over arching umbrella

**Comment by mih** [comment]
The one big technical limitation (that I am aware of) re using git-annex keys as metadata identifiers has been lifted: https://github.com/datalad/datalad/issues/3357#issuecomment-668215466

With this new feature, we can build file keys that use a dataset-specific salt, hence are globally unique, verifiable, but local to the scope of a dataset (i.e. the very same file has two different keys in two datasets). As such, unintended cross-identification of files based on metadata alone can be prevented. This is important when the intent is to communicate anonymous metadata on personal data.

**Comment by mih** [comment]
relevant reference: 

Julie A. McMurry, Nick Juty, Niklas Blomberg, Tony Burdett, Tom Conlin, Nathalie Conte, Mélanie Courtot, John Deck, Michel Dumontier, Donal K. Fellows, Alejandra Gonzalez-Beltran, Philipp Gormanns, Jeffrey Grethe, Janna Hastings, Jean-Karim Hériché, Henning Hermjakob, Jon C. Ison, Rafael C. Jimenez, Simon Jupp, John Kunze, Camille Laibe, Nicolas Le Novère, James Malone, Maria Jesus Martin, Johanna R. McEntyre, Chris Morris, Juha Muilu, Wolfgang Müller, Philippe Rocca-Serra, Susanna-Assunta Sansone, Murat Sariyar, Jacky L. Snoep, Stian Soiland-Reyes, Natalie J. Stanford, Neil Swainston, Nicole Washington, Alan R. Williams, Sarala M. Wimalaratne, Lilly M. Winfree, Katherine Wolstencroft, Carole Goble, Christopher J. Mungall, Melissa A. Haendel, Helen Parkinson. Identifiers for the 21st century: How to design, provision, and reuse persistent identifiers to maximize utility and impact of life science data. PLoS Bio. 2017. https://doi.org/10.1371/journal.pbio.2001414

**Comment by mih** [comment]
> I can have two instances of a dataset locally, cannot identify them uniquely. Might be just needing clarification, that the goal isn't to identify any particular "instance", but rather "substance" ;-)

@yarikoptic I think what you are describing here matches the "distribution-level" description of http://htmlpreview.github.io/?https://github.com/joejimbo/HCLSDatasetDescriptions/blob/master/Overview.html

In the concept/demo I posted in https://github.com/datalad/datalad-metalad/issues/389#issuecomment-1645255207 it would add a `dcat:distribution` property to the top-level objects (which is the version-level description). The property could then have a list of all known locations.

In practice this would be somewhat tricky to implement, because we would need to somehow check, if a particular commit (and/or set of annex keys) is available at that location.

From the perspective of the datalad registry things would be simpler. Pretty much any submission would be a "distribution-level" description (ie. "this is what I have here, locally").
